### PR TITLE
[Dependency Scanning] Attempt to lookup optional transitive dependencies of binary module dependencies

### PR DESCRIPTION
**Pull request by artemcm**
Instead of simply pretending they do not exist, do a best-effort lookup.

### CODE REVIEW

File: test/ScanDependencies/optional_deps_of_binary_testable_imports.swift [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+// RUN: echo "@_implementationOnly import A; public func foo() {}" > %t/Foo.swift
+// REQUIRES: executable_test
+// REQUIRES: objc_interop
+
+@testable import Foo
+
+// Step 1: build a binary swift module for `Foo`, make it testable
+// RUN: %target-swift-frontend -emit-module %t/Foo.swift -emit-module-path %t/Foo.swiftmodule/%target-swiftmodule-name -module-name Foo -I %S/Inputs/CHeaders -I %S/Inputs/Swift -enable-testing
+
+// Step 3: scan dependencies
```
Comment (xymus) [inline test/ScanDependencies/optional_deps_of_binary_testable_imports.swift:13]: No step 2?

File: include/swift/DependencyScan/SerializedModuleDependencyCacheFormat.h [hunk 2]
```diff
@@ -124,6 +124,7 @@ using ModuleInfoLayout =
                    IdentifierIDField,            // moduleName
                    ContextHashIDField,           // contextHash
                    ImportArrayIDField,           // moduleImports
+                   ImportArrayIDField,           // optionalModuleImports
                    DependencyIDArrayIDField      // resolvedDirectModuleDependencies
```
Comment (xymus) [inline include/swift/DependencyScan/SerializedModuleDependencyCacheFormat.h:128]: Does this need a version bump to the format?

**Review by xymus — APPROVED** [review]

File: include/swift/DependencyScan/SerializedModuleDependencyCacheFormat.h (hunk 2 above)
Comment (artemcm) [reply include/swift/DependencyScan/SerializedModuleDependencyCacheFormat.h:128]: Done, thanks!

File: test/ScanDependencies/optional_deps_of_binary_testable_imports.swift (hunk 1 above)
Comment (artemcm) [reply test/ScanDependencies/optional_deps_of_binary_testable_imports.swift]: Sharp eye! Thanks.

**Comment by artemcm** [comment]
@swift-ci smoke test and merge

**Pull request merged**